miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/PoolRestController.java

```diff
@@ -279,6 +283,46 @@ public PoolPickerResponse getPickersBySearch(@RequestParam("platform") String pl
         PaginationFilter.platformType(PlatformType.valueOf(platform)));
   }
 
+  @RequestMapping(value = "avginsertsizes", method = RequestMethod.POST, produces = "application/json")
+  @ResponseBody
+  public JSONObject calculateAverageInsertSizes(@RequestBody JSONObject json) throws IOException {
+    String idString = json.getString("ids");
+    String[] split = idString.split(",");
+    List<Long> ids = new ArrayList<>();
+    for (int i = 0; i < split.length; i++) {
+      ids.add(Long.parseLong(split[i]));
+    }
+    Collection<Pool> pools = poolService.listByIdList(ids);
+    List<String> avgs = new ArrayList<>();
+    for (Pool p : pools) {
+      long sum = 0;
```

Comment:
Use the new Java [stream](https://docs.oracle.com/javase/8/docs/api/java/util/stream/IntStream.html) API:

```
String averageMessage = poolService.get(poolId).getPoolableElementViews().stream()//
   .map(PoolableElementView::getLibraryDnaSize)//
   .filter(Objects::nonNull)//
   .mapToDouble(x -> x)//
   .average()//
   .map(Double::toString).orElse("N/A");
```

Reply:
Hmm, I can't seem to get this to work. I get an issue with the second map():

```
    for (long poolId : ids) {
      String averageMessage = poolService.get(poolId).getPoolableElementViews().stream()//
              .map(PoolableElementView::getLibraryDnaSize)//
              .filter(Objects::nonNull)//
              .mapToDouble(x -> x)//
              .average()//
              .map(Double::toString).orElse("No insert sizes");
          avgs.add(p.getAlias() + ": " + averageMessage);
    }
```
I'm also a bit confused, having never used the Stream API before, how to get the pool alias in the message also?

Reply:
You can store the pool in a temporary variable:
```
Pool p =  poolService.get(poolId);
      String averageMessage = p.getPoolableElementViews().stream()//
              .map(PoolableElementView::getLibraryDnaSize)//
              .filter(Objects::nonNull)//
              .mapToDouble(x -> x)//
              .average()//
              .map(Double::toString).orElse("No insert sizes");
```